src/System.Private.CoreLib/shared/System/Threading/CancellationTokenSource.cs

```diff
@@ -442,7 +446,7 @@ protected virtual void Dispose(bool disposing)
                 // transitioned to and while it's in the NotifyingState.
                 if (_kernelEvent != null)
                 {
-                    ManualResetEvent mre = Interlocked.Exchange(ref _kernelEvent, null);
+                    ManualResetEvent? mre = Interlocked.Exchange<ManualResetEvent?>(ref _kernelEvent!, null);
```

Comment:
@jaredpar, I'm confused by the changes that were necessary on this line.  Here `_kernelEvent` is declared as a `ManualResetEvent?`.  As written here:
```C#
ManualResetEvent? mre = Interlocked.Exchange<ManualResetEvent?>(ref _kernelEvent!, null)
```
it compiles without warning.

However, this:
```C#
ManualResetEvent? mre = Interlocked.Exchange(ref _kernelEvent, null);
```
warns on the `null` argument that `Cannot convert null literal to non-nullable reference or unconstrained type parameter` and hovering over `Exchange` to get IntelliSense for it shows that it inferred the type of `T` to be `ManualResetEvent` rather than `ManualResetEvent?`, even though `_kernelEvent` is `ManualResetEvent?`.  Is it somehow getting messed up by the `_kernelEvent != null` guard above it?

Similarly, if I change it to be:
```C#
ManualResetEvent? mre = Interlocked.Exchange<ManualResetEvent?>(ref _kernelEvent, null);
```
that warns on the `_kernelEvent` argument "Possible null reference assignment", which should be fine, given that the type of `T` here is nullable.

Is this by design or a bug?

Reply:
cc: @gafter @cston

Reply:
Yes, the null check is causing `ref _kernelEvent` to be inferred as non-nullable.